### DISCUSSION TITLE
Select with boolean values not working with actual boolean

### DISCUSSION THREAD

**Discussion by mokhosh** [livewire/livewire#4356] [7. Bugs (archived)] [answered]
I have a boolean select component like this:
```blade
<select {{ $attributes }}>
    <option value="0">No</option>
    <option value="1">Yes</option>
</select>
```
This works perfectly with 0s and 1s, but if I have `$casts` and the attribute is being cast to a `boolean` it won't work. How can I make it work with both actual boolean values and 0s and 1s?

**Comment by joshhanley** [comment]
@mokhosh can't you just set the values to `true` and `false`?

If not then you'd need to make use of an `updating` hook with value passed by reference to cast the truthy/falsey strings to a boolean before it gets assigned to the attribute.

```php
public function updatingPropertyName(&$value)
{
    $value = filter_var($value, FILTER_VALIDATE_BOOLEAN);
}
```

Hope this helps!

**Reply by mokhosh** [reply]
That's a problem with html I guess. It doesn't recognise `true` and `false` as boolean values. It treats them as strings and they either create bugs or exceptions, because they're both converted to `true`.

I can use hooks in all of my components, but that's really tedious. I wish there was a more elegant way for such a fundamental problem.

**Reply by joshhanley** [reply]
@mokhosh yeah agreed.

It'd be interesting to know how Laravel handles converting strings into booleans when getting data from a request and saving it to a boolean property, as it must do the conversion at some point (maybe it doesn't)?

If it does maybe we can setup a similar process?

**Reply by mokhosh** [reply]
I don't think Laravel converts data, but this problem won't come up in Laravel because it can only happen when you have data binding. In other situations you will either not have anything or have the explicit value in your blade.

**Reply by ju5t** [reply]
@joshhanley I think this is a missing feature in Livewire. We switched some boolean options to radio buttons and they do bind correctly. Looks like it's specific to `select`. I prefer to use a `select` over radio buttons, but at the moment it's unsupported?

**Reply by joshhanley** [reply]
@ju5t interesting! Yeah agreed, if one works, then the other should. Can you share a sample component that has both examples so I can compare them? Thanks!

**Reply by ju5t** [reply]
@joshhanley do you prefer a full Livewire project or a gist? Either is fine by me. 

And while we're at it 😄 I'd also like `null` to be treated as empty; so `<option value disabled>Select something...</option>` would be a valid option if the wired property is `null`. Right now, `null` is somehow treated as a string which is a bit odd.

I can add an example for this one too.

**Reply by joshhanley** [reply]
@ju5t sorry for the delay, a gist is fine as long as I can just copy the component into my test project to get it running quickly 🙂

`null` is a hard one. Technically everything from html is a string, so an option with an empty value is actually just an empty string.
Laravel get's around this by using a `convertEmptyStringsToNull` middleware. But this can actually break Livewire if it were applied, due to it changing the checksum, causing a corruption error.

I've got some solutions as it what can happen for null but it just depends on what is demoed in V3 at Laracon. So will wait for that to see first.

In the meantime I just use a custom trait. It's not perfect but I think it's enough

```php
trait ConvertEmptyStringsToNull
{
    public function updatedConvertEmptyStringsToNull($name)
    {
        $value = $this->getPropertyValue($name);

        if (is_string($value) && $value === '') {
            // This method is a macro, see AppServiceProvider
            $this->setPropertyValue($name, null);
        }
    }
}
```

Hope this helps!

**Comment by lucadegasperi** [comment]
This is still an ongoing issue I encounter daily and one of the most frustrating aspects of using Livewire. I just wish "0","1","true", "false" to be treated as boolean values. I find myself disabling the model casting to prevent the issue which is a bummer.

**Comment by pxlrbt** [comment]
Was there any update to this?

**Reply by lucadegasperi** [reply, marked as answer]
livewire v3 solves this with the wire:model.boolean attribute modifier.

**Reply by pxlrbt** [reply]
Nice. Thanks for the info!
